### PR TITLE
Feat/add validation rules

### LINKED ISSUE #141: feat: add validation rules in `Memory` structure
Currently, there are two essential utility functions absent in the `Memory` structure that play a crucial role in validating memory. We need the following implementations in Ziglang:

### A `validateMemoryCell` method
- Purpose: To validate a specific memory cell.
- Link to Rust Implementation: [`validate_memory_cell`](https://github.com/lambdaclass/cairo-vm/blob/e6171d66a64146acc16d5512766ae91ae044f297/vm/src/vm/vm_memory/memory.rs#L315-L327)

### A `validateExistingMemory` method
- Purpose: To validate the entire current memory based on predefined validation rules.
- Link to Rust Implementation: [`validate_existing_memory`](https://github.com/lambdaclass/cairo-vm/blob/e6171d66a64146acc16d5512766ae91ae044f297/vm/src/vm/vm_memory/memory.rs#L330-L347)

### PR DESCRIPTION
**Pull request by oxlime**
Will need #150 to be merged first. Closes #141 

There is some changes that need to be made in range check that I can tackle here or put into another pr if better. As far as this issue and the memory file this should be working. 

Let me know if more tests needed.

### CODE REVIEW

**Comment by tcoratger** [comment]
@oxlime #150 is merged :)

File: src/vm/memory/memory.zig [hunk 1]
```diff
@@ -326,6 +342,24 @@ fn setUpMemory(memory: *Memory, comptime vals: anytype) !void {
     }
 }
 
+test "Memory: validate memory cell" {
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:345]: I think that we should add two more tests here:
- One to assert that if the segment index of the provided address in `validateMemoryCell` does not exist in `validation_rules`, then nothing change.
- One to assert that if the segment index of the provided address in `validateMemoryCell` exist in `validation_rules` but the address already exists in `validated_addresses` then nothing change.

**Comment by oxlime** [comment]
> @oxlime #150 is merged :)

awesome will get this done in next 24hrs

File: src/vm/memory/memory.zig (hunk 1 above)
Comment (oxlime) [reply src/vm/memory/memory.zig:345]: Gotcha. Also i have to finish the validate existing memory function and will add more tests to that as well.

File: src/vm/memory/memory.zig [hunk 2]
```diff
@@ -399,6 +427,124 @@ pub fn setUpMemory(memory: *Memory, allocator: Allocator, comptime vals: anytype
...
+
+    try expect(
+        segments.memory.validated_addresses.contains(Relocatable.new(0, 2)),
+    );
+    try expect(
+        segments.memory.validated_addresses.contains(Relocatable.new(0, 5)),
+    );
+    try expect(
+        segments.memory.validated_addresses.contains(Relocatable.new(0, 7)),
+    );
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:466]: Here can we also verify that the other addresses are not in `validated_addresses` because you only added the 0 segment as a validation rule?

File: src/vm/memory/memory.zig [hunk 3]
```diff
@@ -359,6 +360,33 @@ pub const Memory = struct {
...
+    /// # Arguments
+    /// - `address`: The source Relocatable address of the memory cell to be checked.
+    ///
+    /// # Returns
+    /// This function returns an error if the validation fails due to invalid conditions.
+    pub fn validateMemoryCell(self: *Self, address: Relocatable) !void {
+        if (self.validation_rules.get(@intCast(address.segment_index))) |rule| {
+            if (!self.validated_addresses.contains(address)) {
+                var list = rule(self, address);
+                try self.validated_addresses.addAddresses(list.?);
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:375]: Here what happens if `list` is `null`? Because you try to access the element inside but if `null`, this will raise an error no?

File: src/vm/memory/memory.zig [hunk 4]
```diff
@@ -399,6 +427,124 @@ pub fn setUpMemory(memory: *Memory, allocator: Allocator, comptime vals: anytype
...
+    var segments = try MemorySegmentManager.init(allocator);
+    defer segments.deinit();
+
+    var builtin = RangeCheckBuiltinRunner.new(8, 8, true);
+    builtin.initializeSegments(segments);
+    try builtin.addValidationRule(segments.memory);
+    var seg = segments.addSegment();
+    _ = seg;
+
+    try segments.memory.set(std.testing.allocator, Relocatable.new(0, 1), relocatable.fromFelt(starknet_felt.Felt252.one()));
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:481]: Maybe using here a format like:
```zig
 try setUpMemory(
    segments.memory,
    std.testing.allocator,
    .{.{ .{ 0, 1 }, .{ 1 } }},
);
defer segments.memory.deinitData(std.testing.allocator);
```

Related https://github.com/keep-starknet-strange/cairo-zig/issues/170

File: src/vm/memory/memory.zig [hunk 5]
```diff
@@ -399,6 +427,124 @@ pub fn setUpMemory(memory: *Memory, allocator: Allocator, comptime vals: anytype
...
+    var segments = try MemorySegmentManager.init(allocator);
+    defer segments.deinit();
+
+    var builtin = RangeCheckBuiltinRunner.new(8, 8, true);
+    builtin.initializeSegments(segments);
+
+    var seg = segments.addSegment();
+    _ = seg;
+
+    try segments.memory.set(std.testing.allocator, Relocatable.new(0, 1), relocatable.fromFelt(starknet_felt.Felt252.one()));
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:504]: Same here, maybe using:
```zig
 try setUpMemory(
    segments.memory,
    std.testing.allocator,
    .{.{ .{ 0, 1 }, .{ 1 } }},
);
defer segments.memory.deinitData(std.testing.allocator);
```

File: src/vm/memory/memory.zig [hunk 6]
```diff
@@ -399,6 +427,124 @@ pub fn setUpMemory(memory: *Memory, allocator: Allocator, comptime vals: anytype
...
+    defer segments.deinit();
+
+    var builtin = RangeCheckBuiltinRunner.new(8, 8, true);
+    builtin.initializeSegments(segments);
+    try builtin.addValidationRule(segments.memory);
+
+    var seg = segments.addSegment();
+    _ = seg;
+
+    try segments.memory.set(std.testing.allocator, Relocatable.new(0, 1), relocatable.fromFelt(starknet_felt.Felt252.one()));
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:528]: Same for the format

File: src/vm/memory/memory.zig [hunk 7]
```diff
@@ -399,6 +427,124 @@ pub fn setUpMemory(memory: *Memory, allocator: Allocator, comptime vals: anytype
...
+    );
+    try expect(
+        segments.memory.validated_addresses.contains(Relocatable.new(0, 5)),
+    );
+    try expect(
+        segments.memory.validated_addresses.contains(Relocatable.new(0, 7)),
+    );
+}
+
+test "Memory: validate memory cell" {
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:469]: I would also add a test for the `null` case, meaning:

```zig
 try setUpMemory(
    segments.memory,
    std.testing.allocator,
    .{.{ .{ 0, 1 }, .{ 1 } }, .{ .{ 0, 10 }, .{ 1 } } },
);
defer segments.memory.deinitData(std.testing.allocator);
```

and 

```zig
try segments.memory.validateMemoryCell(Relocatable.new(0, 7));
```

File: src/vm/builtins/builtin_runner/range_check.zig [hunk 8]
```diff
@@ -227,19 +227,23 @@ pub const RangeCheckBuiltinRunner = struct {
...
             return null;
         }).tryIntoFelt() catch {
-            return RunnerError.BuiltinExpectedInteger;
+            return null;
         };
 
-        if (num.Mask <= N_PARTS * INNER_RC_BOUND_SHIFT) {
-            return try result.append(address);
+        if (@bitSizeOf(u256) - @clz(num.toInteger()) <= N_PARTS * INNER_RC_BOUND_SHIFT) {
+            //try result.append(address);
```
Comment (tcoratger) [inline src/vm/builtins/builtin_runner/range_check.zig:241]: Comments can be removed here no?

File: src/vm/builtins/builtin_runner/range_check.zig [hunk 9]
```diff
@@ -227,19 +227,23 @@ pub const RangeCheckBuiltinRunner = struct {
     ///
     /// An `ArrayList(Relocatable)` containing the rules address
     /// verification fails.
-    pub fn rangeCheckValidationRule(memory: *Memory, address: Relocatable, allocator: Allocator) !std.ArrayList(Relocatable) {
-        var result = ArrayList(Relocatable).init(allocator);
-        errdefer result.deinit();
+    pub fn rangeCheckValidationRule(memory: *Memory, address: Relocatable) ?[]const Relocatable {
+        // pub fn rangeCheckValidationRule(memory: *Memory, address: Relocatable, allocator: Allocator) !std.ArrayList(Relocatable) {
```
Comment (tcoratger) [inline src/vm/builtins/builtin_runner/range_check.zig:231]: Comments can be removed here no?

File: src/vm/memory/memory.zig [hunk 10]
```diff
@@ -19,7 +19,8 @@ const MemorySegmentManager = @import("./segments.zig").MemorySegmentManager;
 const RangeCheckBuiltinRunner = @import("../builtins/builtin_runner/range_check.zig").RangeCheckBuiltinRunner;
 
 // Function that validates a memory address and returns a list of validated adresses
-pub const validation_rule = *const fn (*Memory, Relocatable) std.ArrayList(Relocatable);
+//pub const validation_rule = *const fn (*Memory, Relocatable) std.ArrayList(Relocatable);
```
Comment (tcoratger) [inline src/vm/memory/memory.zig:22]: This comment can be removed here no?

**Review by tcoratger — APPROVED** [review]
